### PR TITLE
Revise code around record/replay of ScriptContext enter/exit.

### PR DESCRIPTION
**Pull request by mrkmarron**
- Refactored code to ensure script context enter/exit is captured accurately in record.
- Refactored replay code to ensure we enter/exit contexts exactly as in record.
- Updated exception propagation code to ensure replay fidelity and needed interaction with time-travel operations.

### CODE REVIEW

**Comment by mrkmarron** [comment]
@agarwal-sandeep please take a look.

File: lib/Jsrt/Jsrt.cpp [hunk 1]
```diff
@@ -2152,19 +2146,16 @@ CHAKRA_API JsCallFunction(_In_ JsValueRef function, _In_reads_(cargs) JsValueRef
         Js::Arguments jsArgs(callInfo, reinterpret_cast<Js::Var *>(args));
 
 #if ENABLE_TTD
-        double ttdStartTime = -1.0;
-
+        TTD::TTDJsRTActionResultAutoRecorder _actionEntryPopper(TTD::ContextWrapperEnterExitStatus::EnterContextAPIWrapper);
         if(PERFORM_JSRT_TTD_RECORD_ACTION_CHECK(scriptContext))
```
Comment (agarwal-sandeep) [inline lib/Jsrt/Jsrt.cpp]: Should this be inside the if?

File: lib/Jsrt/Jsrt.cpp [hunk 2]
```diff
@@ -3491,36 +3468,44 @@ CHAKRA_API JsTTDStartTimeTravelRecording()
...
-        if (threadContext->TTDLog == nullptr)
-        {
-            AssertMsg(false, "Need to create in TTD mode.");
-            return JsErrorCategoryUsage;
-        }
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    JsErrorCode cCheck = CheckContext(currentContext, true);
+    if(cCheck != JsNoError)
+    {
+        AssertMsg(false, "This shouldn't happen!!!");
```
Comment (agarwal-sandeep) [inline lib/Jsrt/Jsrt.cpp]: >AssertMsg(false, "This shouldn't happen!!!"); [](start = 8, length = 45)

nit: Seems like not a very useful assert given that this is external API and can be called when this is supposed to fail like when context is not set.

**Comment by agarwal-sandeep** [comment]
:clock1:

**Pull request merged**